interop-tests/src/test/scala/akka/grpc/scaladsl/LoadBalancingIntegrationSpec.scala

```diff
@@ -139,23 +138,23 @@ class LoadBalancingIntegrationSpec extends AnyWordSpec with Matchers with Before
       service.greetings.get should be(100)
     }
 
-    "eventually fail when no valid endpoints are provided" in {
-      val discovery =
-        new MutableServiceDiscovery(
-          List(new InetSocketAddress("example.invalid", 80), new InetSocketAddress("example.invalid", 80)))
-      val client = GreeterServiceClient(
-        GrpcClientSettings
-          .usingServiceDiscovery("greeter", discovery)
-          .withTls(false)
-          .withGrpcLoadBalancingType("round_robin")
-          // Low value to speed up the test
-          .withConnectionAttempts(2))
-
-      val failure =
-        client.sayHello(HelloRequest(s"Hello friend")).failed.futureValue.asInstanceOf[StatusRuntimeException]
-      failure.getStatus.getCode should be(Code.UNAVAILABLE)
-      client.closed.failed.futureValue shouldBe a[ClientConnectionException]
```

Comment:
Removing this line fixes the test. 
`grpc-java` `1.29.0` changed the way subchannel events are notified (in load-balanced situations). With that change, the `ChannelUtils` listener for channel changes no longer works when using load balanced channels because it can no longer count reconnection attempts. As a consequence, a scenario like the one in the test will no longer fail but retry indefinitely.

Once I realized the reconnection would work indefinitely I decided the test could be removed.

And, if we remove the test, and then start pulling the thread, we should probably deprecate `withConnectionAttempts(2)` (or maybe re-implement the underlying logic if we consider a `connectionAttempts` counter makes any sense now that we are using the internal `NameResolver` engine provided by `groc-java`.

Before pushing forward and deleting all usages of `connectionAttempts` I preferred raising this comment for discussion.

Reply:
I guess it makes sense to follow the lead of the reference implementation and just keep retrying when loadbalancing, then.

> if we remove the test, and then start pulling the thread, we should probably deprecate withConnectionAttempts(2)

Isn't it used in the non-loadbalanced scenario either?

Reply:
> Isn't it used in the non-loadbalanced scenario either?

Yeah, I had the same thought over breakfast earlier today. I think we don't have a test for that and, even if we have such a test, we should improve documentation or API around `withConnectionAttempts`.

My main concern is whether we should improve the load-balanced use-cases. For example:

 - in situations where none of the records returned by the service discovery are valid, I'd like to see the aggregated channel to fail eventually (like it does in non-balanced usages). Or, at least, I'd like to see failure traces in the logs (I think we can have a look at logging settings for that).

One of the reasons this took me a while to diagnose was that the test code (aka user code) had zero visibility over the underlying failures. Also, I (the developer) didn't see any apparent failure or retries happening even with `DEBUG` logs.

Reply:
> in situations where none of the records returned by the service discovery are valid, I'd like to see the aggregated channel to fail eventually

I agree. If we don't have any insight in the progress of reconnecting / number of reconnects, I guess the best we could do is wait for a timeout: if a channel remains in TRANSIENT_FAILURE for X amount of time, assume it won't ever successfully connect and fail.

Does the reference implementation have any such timeout?

Reply:
> Isn't it used in the non-loadbalanced scenario either?

Yeah, I had the same thought over breakfast earlier today. I think we don't have a test for that and, even if we have such a test, we should improve documentation or API around `withConnectionAttempts`.

------------------------------------

I had the answer above DRAFTed and then I realized that I don't know how to control wether the `Channel` provided is a single-Channel or a load-balanced one with multiple sub-channels.

Maybe we're no longer using a non LoadBalanced client. The client factory methods all create a `ServiceDiscovery` instance (maybe `akka-dns`, maybe `config`, maybe other) and, since we always inject a `NamedResolver` based on that `ServiceDiscovery` it's possible we're always running a `Channel` with sub-channel(s).

I need to dig that deeper.

Reply:
The upstream default is `pick_first` (no loadbalancing, https://github.com/grpc/grpc-java/blob/c91e127f93f1a60e5625c8bb71c4e09f7858966b/core/src/main/java/io/grpc/internal/GrpcUtil.java#L245), which we only override if 'grpc-load-balancing' is set (https://github.com/akka/akka-grpc/blob/master/runtime/src/main/scala/akka/grpc/internal/NettyClientUtils.scala#L64), which is not set by default (https://github.com/akka/akka-grpc/blob/master/runtime/src/main/resources/reference.conf#L25)

Reply:
(I had multiple GH windows and posted 2 variants of the same comment)

Ok, so with `grpc-java` `1.29.0` we should keep `withConnectionAttempts` around but it'll only be used when `withGrpcLoadBalancingType` is not invoked. I've just tested that and it works.

I think we have to improve documentation with multiple things:

1. when `withGrpcLoadBalancingType` is used, `withConnectionAttempts` is ignored, the underlying (`akka-grpc`) implementation for backoff reconnecting will be used indefintely and `client.closed` only completes when `client.close()` is directly invoked. Note: TIL the load balancing in the `akka-grpc` implementation is per call.
2. when `withGrpcLoadBalancingType` is not used, but a `ServiceDiscovery` is provided alls calls are sent to a single server instance but users can map over `client.closed` to reconnect again which will trigger a new `ServiceDiscovery` lookup.

Reply:
> Does the reference implementation have any such timeout?

I'm reading docs here and there trying to see what other choices we have.

> The upstream default is pick_first (no loadbalancing...

Again! 🤦🏼‍♂️ I should refresh the page before answering...

Reply:
> the underlying (`akka-grpc`) implementation for backoff reconnecting will be used

Do you mean `grpc-java`?

> TIL the load balancing in the akka-grpc implementation is per call

What do you mean?

> all calls are sent to a single server instance but users can map over client.closed to reconnect again which will trigger a new ServiceDiscovery lookup.

I would *expect* that when that single server instance fails, grpc-java will automatically fail over to another instance (triggering a new ServiceDiscovery lookup) as needed. Do you think we need to do something extra there?

In any case would be good to add tests for those scenario's.

Reply:
(🤦🏼‍♂️ Today is not my day)

In both statements above I meant `grpc-java` (not `akka.grpc`).

> TIL the load balancing in the ~akka-grpc~ _grpc-java_ implementation is per call

🤦🏼‍♂️ It is `round-robin`! What did I expect?

Reply:
> I would _expect_ that when that single server instance fails, grpc-java will automatically fail over to another instance (triggering a new ServiceDiscovery lookup) as needed. Do you think we need to do something extra there?

I think our code listening on the number of re-connection attempts is the issue here and we may want to remove it once and for all. I mean, having moved to an implementation based on `NameResolver`, even when using `pick-first` you are probably correct and `grpjc-java` is doing a new name resolution. So, it doesn't make sense that close after a certain [number](https://github.com/akka/akka-grpc/blob/88252782b64809d3d44d7510f2e648c13aa5aa96/runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala#L60-L77) of reconnection attempts. I mean, we could keep the code around for logging purposes (even though that code is useless in `round-robin`) but we probably can drop the `withConnectionAttempts` setting and `closed` future.

> In any case would be good to add tests for those scenario's.

+1